File: ruby/sorting/mergesort.rb
def mergsort(arry)
  return arry if arry.length < 2
  length = arry.length
  mid = (length/2).floor
  left = arry[0 ... mid]
  right = arry[mid .. -1]
  return glue(mergsort(left), mergsort(right))
end

def glue(arry1, arry2)
  glued = []
  while arry1.length > 0  && arry2.length > 0
    if arry1[0] <= arry2[0]
      glued << arry1.shift
    elsif arry1[0] >= arry2[0]
      glued << arry2.shift
    end
  end
  (glued << arry1 << arry2).flatten
end

p mergsort([3,2,54,2,1])

File: js/sorting/mergesort.js
function mergsort(arry) {
  if(arry.length < 2) return arry;
  var length = arry.length;
  var mid = Math.floor(length/2);
  var left = arry.slice(0, mid);
  var right = arry.slice(mid, length)
  return glue(mergsort(left), mergsort(right))
}

function glue(arry1, arry2) {
  var sorted = [];
  while(arry1.length && arry2.length) {
    if(arry1[0] <= arry2[0]) {
      sorted.push(arry1.shift());
    } else {
      sorted.push(arry2.shift());
    }
  }
  return sorted.concat(arry1, arry2)
}


console.log(mergsort([100,6,2,5,3,2,1,35,2,5,1]))
